File: src/office/metadata/formulaMetadata.ts
export interface VisualTeXFormulaMetadata {
  schema: "visualtex-formula";
  schemaVersion: 1;

  formulaId: string;

  title: string;
  latex: string;

  lines: Array<{
    id: string;
    latex: string;
  }>;

  codeFormat: string;
  displayMode: "inline" | "block";

  createdWithVersion: string;
  updatedWithVersion: string;

  createdAt: string;
  updatedAt: string;
}

export const VISUALTEX_FORMULA_SCHEMA = "visualtex-formula" as const;
export const VISUALTEX_FORMULA_SCHEMA_VERSION = 1 as const;

export function isVisualTeXFormulaMetadata(
  value: unknown,
): value is VisualTeXFormulaMetadata {
  if (!value || typeof value !== "object") return false;
  const candidate = value as Partial<VisualTeXFormulaMetadata>;
  return (
    candidate.schema === VISUALTEX_FORMULA_SCHEMA &&
    candidate.schemaVersion === VISUALTEX_FORMULA_SCHEMA_VERSION &&
    typeof candidate.formulaId === "string" &&
    typeof candidate.latex === "string" &&
    Array.isArray(candidate.lines)
  );
}

File: src-windows/VisualTeX.WindowsOffice.Contracts/FormulaMetadata.cs
using System;
using System.Collections.Generic;
using System.Text.Json.Serialization;

namespace VisualTeX.WindowsOffice.Contracts;

public sealed class FormulaMetadata
{
    [JsonPropertyName("schema")]
    public string Schema { get; set; } = "visualtex-formula";

    [JsonPropertyName("schemaVersion")]
    public int SchemaVersion { get; set; } = 1;

    [JsonPropertyName("formulaId")]
    public string FormulaId { get; set; } = string.Empty;

    [JsonPropertyName("title")]
    public string Title { get; set; } = string.Empty;

    [JsonPropertyName("latex")]
    public string Latex { get; set; } = string.Empty;

    [JsonPropertyName("lines")]
    public List<FormulaLine> Lines { get; set; } = new();

    [JsonPropertyName("codeFormat")]
    public string CodeFormat { get; set; } = string.Empty;

    [JsonPropertyName("displayMode")]
    public string DisplayMode { get; set; } = "block";

    [JsonPropertyName("numbered")]
    public bool Numbered { get; set; }

    [JsonPropertyName("createdWithVersion")]
    public string CreatedWithVersion { get; set; } = string.Empty;

    [JsonPropertyName("updatedWithVersion")]
    public string UpdatedWithVersion { get; set; } = string.Empty;

    [JsonPropertyName("createdAt")]
    public string CreatedAt { get; set; } = string.Empty;

    [JsonPropertyName("updatedAt")]
    public string UpdatedAt { get; set; } = string.Empty;

    public void Validate()
    {
        if (Schema != "visualtex-formula" || SchemaVersion != 1)
            throw new InvalidOperationException("Unsupported VisualTeX formula metadata schema.");
        if (!Guid.TryParse(FormulaId, out _))
            throw new InvalidOperationException("VisualTeX formulaId must be a UUID.");
        if (Lines.Count == 0)
            throw new InvalidOperationException("VisualTeX formula metadata requires at least one line.");
        if (Numbered && !string.Equals(DisplayMode, "block", StringComparison.Ordinal))
            throw new InvalidOperationException("Only display formulas can use equation numbering.");
    }
}

public sealed class FormulaLine
{
    [JsonPropertyName("id")]
    public string Id { get; set; } = string.Empty;

    [JsonPropertyName("latex")]
    public string Latex { get; set; } = string.Empty;
}
